### PR TITLE
conv2d primitive support from backends; support for ‘torch.nn.functional.conv2d’ op of pytorch frontend

### PR DESCRIPTION
**Pull request by ethancaballero**
- this includes full forward and backward for all input variations
- currently only supports PyTorch backend
- `await force_pending(engine.check(typ, type_token(arg), typ))` in line 95 of myia/abstract/prim.py 
- @abergeron will do relay backend support before PR is accepted (nnvm backend won't be needed once Arnaud's Backend reorganization (https://github.com/mila-iqia/myia/pull/238) PR is done)

### CODE REVIEW

File: myia/abstract/prim.py [hunk 1]
```diff
@@ -880,6 +890,63 @@ class _GetAttrInferrer(Inferrer):
     return type(a)(a.element, {SHAPE: c_shp})
 
 
+def _pair(self, x):
+    if isinstance(x, AbstractScalar):
+        x = self.require_constant(x, argnum=0)
+        return tuple(repeat(x, 2))
```
Comment (breuleux) [inline myia/abstract/prim.py]:
```suggestion
        return (x, x)
```

File: myia/abstract/prim.py [hunk 2]
```diff
@@ -880,6 +890,63 @@ class _GetAttrInferrer(Inferrer):
...
+    elif isinstance(x, AbstractTuple):
+        return tuple(self.require_constant(e, argnum=0) for e in x.elements)
```
Comment (breuleux) [inline myia/abstract/prim.py]: Check that the length is 2 and raise a `MyiaTypeError` if it's not.

File: myia/compile/backends/pytorch.py [hunk 3]
```diff
@@ -148,10 +149,115 @@ def _impl(array):
...
+                           dilation, groups):
+    """Wrap of conv2d_input_grad for pytorch."""
+    input_size = tuple(i.item() for i in input_size)
+
+    if isinstance(stride, np.ndarray):
+        stride = stride.item()
+    elif isinstance(stride, tuple):
+        stride = tuple(s.item() for s in stride)
+    else:
+        pass
```
Comment (breuleux) [inline myia/compile/backends/pytorch.py]: Can you make a helper function for this? I can see 6 instances of this pattern.

File: myia/frontends/pytorch.py [hunk 4]
```diff
@@ -91,6 +104,9 @@ def _array_zero(xs):
     'relu': relu,
     'reshape': P.reshape,
     'sigmoid': sigmoid,
+
+    'shape': property(P.shape),
+
```
Comment (breuleux) [inline myia/frontends/pytorch.py]: Fix spacing.

File: myia/frontends/pytorch_functions.py [hunk 5]
```diff
@@ -29,6 +29,27 @@
...
+    if padding_mode != 'zeros':
+        raise Exception("'zeros' is the only padding_mode that is currently
+                        supported.")
+    #"""
+
+    if bias is not None:
+        ret = P.conv2d(input, weight, stride, padding, dilation, groups) \
+            + bias.reshape((1, bias.shape[0], 1, 1))
+    else:
+        ret = P.conv2d(input, weight, stride, padding, dilation, groups)
```
Comment (breuleux) [inline myia/frontends/pytorch_functions.py]: It'd be a bit better to pull `P.conv2d(input, weight, stride, padding, dilation, groups)` out of the branches.

File: myia/prim/py_implementations.py [hunk 6]
```diff
@@ -470,6 +470,29 @@ def dot(a, b):
     return np.dot(a, b)
 
 
+@register(primops.conv2d)
+def conv2d(input, weight, stride, padding, dilation, groups):
+    """Implement 2d_convolution."""
+    # TODO: this return is wrong
```
Comment (breuleux) [inline myia/prim/py_implementations.py]: This is self-evident :)

File: myia/validate.py [hunk 7]
```diff
@@ -125,6 +125,11 @@ def validate_abstract(self, a: (type(None), AbstractExternal), uses):
     P.casttag,
     P.tagged,
     P.unsafe_static_cast,
+
+    P.conv2d,
+
```
Comment (breuleux) [inline myia/validate.py]: Fix spacing.

File: tests/frontends/test_pytorch.py [hunk 8]
```diff
@@ -661,3 +657,347 @@ def step(model, inp):
...
+                                     requires_grad=True))
+    weight = nn.Parameter(torch.randn(3, 1, 3, 3,
+                                      dtype=getattr(torch, args.dtype),
+                                      requires_grad=True))
+
+    bias = nn.Parameter(torch.randn(3, dtype=getattr(torch, args.dtype),
+                                    requires_grad=True))
+
+    class Conv2dMod(nn.Module):
+        def __init__(self, ):
```
Comment (breuleux) [inline tests/frontends/test_pytorch.py]:
```suggestion
        def __init__(self):
```

File: tests/test_infer.py [hunk 9]
```diff
@@ -2277,3 +2277,8 @@ def test_is(x, y):
 )
 def test_is_not(x, y):
     return x is not y
+
+
+def test_pair():
+    with pytest.raises(MyiaTypeError):
+        _pair(None, None)
```
Comment (breuleux) [inline tests/test_infer.py]: This test would be more appropriate as a test case in `test_infer` where you pass an illegal type, like an array.
